src/vivarium/examples/boids/forces.py

```diff
@@ -0,0 +1,168 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+import pandas as pd
+
+from vivarium import Component
+from vivarium.framework.engine import Builder
+from vivarium.framework.population import SimulantData
+
+
+class Force(Component, ABC):
+    ##############
+    # Properties #
+    ##############
+    @property
+    def configuration_defaults(self) -> Dict[str, Any]:
+        return {
+            self.__class__.__name__.lower(): {
+                "max_force": 0.03,
+            },
+        }
+
+    columns_required = ["x", "y", "vx", "vy"]
+
+    #####################
+    # Lifecycle methods #
+    #####################
+
+    def setup(self, builder: Builder) -> None:
+        self.config = builder.configuration[self.__class__.__name__.lower()]
+        self.max_velocity = builder.configuration.movement.max_velocity
+
+        self.neighbors = builder.value.get_value("neighbors")
+
+        builder.value.register_value_modifier(
+            "acceleration",
+            modifier=self.apply_force,
+        )
+
+    ##################################
+    # Pipeline sources and modifiers #
+    ##################################
+
+    def apply_force(self, index: pd.Index, acceleration: pd.DataFrame) -> pd.DataFrame:
+        neighbors = self.neighbors(index)
+        pop = self.population_view.get(index)
+        pairs = self._get_pairs(neighbors, pop)
+
+        raw_force = self.calculate_force(pairs)
+        force = self._normalize_and_limit_force(
+            force=raw_force,
+            velocity=pop[["vx", "vy"]],
+            max_force=self.config.max_force,
+            max_velocity=self.max_velocity,
+        )
+
+        acceleration.loc[force.index] += force[["x", "y"]]
+        return acceleration
+
+    ##################
+    # Helper methods #
+    ##################
+
+    @abstractmethod
+    def calculate_force(self, neighbors: pd.DataFrame):
+        pass
+
+    def _get_pairs(self, neighbors: pd.Series, pop: pd.DataFrame):
+        pairs = (
+            pop.join(neighbors.rename("neighbors"))
+            .reset_index()
+            .explode("neighbors")
+            .merge(
+                pop.reset_index(),
+                left_on="neighbors",
+                right_index=True,
+                suffixes=("_self", "_other"),
+            )
+        )
+        pairs["distance_x"] = pairs.x_other - pairs.x_self
+        pairs["distance_y"] = pairs.y_other - pairs.y_self
+        pairs["distance"] = np.sqrt(pairs.distance_x**2 + pairs.distance_y**2)
+
+        return pairs
+
+    def _normalize_and_limit_force(
+        self,
+        force: pd.DataFrame,
+        velocity: pd.DataFrame,
+        max_force: float,
+        max_velocity: float,
+    ):
+        normalization_factor = np.where(
```

Comment:
I don't think I totally understand what's going on with this normalization. It looks like it is doing: 
$$\vec{F} = \vec{F} * \frac{v_{max}}{\left| \vec{F} \right|} - \vec{v} = \hat{F} *v_{max} -\vec{v} $$

Which will ensure that the resulting velocity after application will have a magnitude $v_{max}$ (that is, if the max force scaling were not applied), regardless of the calculated magnitude of the force, right? Is that what you want?

Reply:
Note for posterity: this is not super obvious. The direction of the "steering force" depends on the relative magnitudes of the speed and max_speed, but the max_force scaling below should almost always determine the magnitude of the steering force. See https://www.red3d.com/cwr/papers/1999/gdc99steer.pdf for the background.

Overall, we think this is probably right.